display.py

```diff
@@ -21,271 +24,309 @@ def font(family='', bold=False, italic=False):
     # enable stdout
     sys.stdout = oldstdout
 
-# TODO use cairo context transforms properly instead of passing scale around everywhere
-def transform_x(x, scale):
-    return scale/2 + x * scale
+class BaseDisplay:
+    def __init__(self, cell_fn=None, edge_fn=None, point_fn=None, padding=0.5):
+        self.cell_fns = defaultdict(lambda: cell_fn)
+        self.edge_fns = defaultdict(lambda: edge_fn)
+        self.point_fns = defaultdict(lambda: point_fn)
```

Comment:
Why all this `defaultdict` and `set_*_fn` machinery? The old `draw_grid` machinery was much simpler; since you're already making a subclass of `BaseDisplay` for each grid type, you could give those types specific members for each of those grids' edge types. What value is this abstraction earning?

(I don't know why these `Display` classes have to be classes at all, actually; users only ever create them and then call a method on them. But I'll assume temporarily that you have a good reason I just haven't seen yet, and you aren't just succumbing to OOP-mania.)

Reply:
I had a notion that you would sometimes want to display multiple grids consecutively, but basically in the same way. But I guess that didn't really happen. So I agree that the separation of the two calls doesn't really seem necessary, unless you want to set individual type-based drawing functions.

The `set_*_fn` machinery is now ALMOST never used, because we almost never have to do different things for vert vs. horiz edges. That was one of the goals. I think the only solver that actually uses it is sudoku, to figure out which edges should be bolded. I could have fixed that too with an incredibly opaque manipulation on `edge.vector` but I was pretty sure you would have yelled at me.

Reply:
Ha ha yeah, I'm not arguing for more opacity. I'm just saying we went from a function with parameters to a class hierarchy with getters and setters backed by defaultdicts whose keys are things that only the subclasses know about, and when I see that pattern I wonder why we're too good for:

```python
class WibbleDisplay:
  def __init__(self, foo=None, foo_special1=None, foo_special2=None):
    self.foo_special1 = foo_special1 or foo
    self.foo_special2 = foo_special2 or foo
```

or, if OOP can be avoided:

```python
def display_wibble(foo=None, foo_special1=None, foo_special2=None):
  if foo_special1 is None:
    foo_special1 = foo
  if foo_special2 is None:
    foo_special2 = foo
  # do work here
```

because fields or local variables are way simpler than schemes involving defaultdicts and accessor pairs!